lib/AST/ASTPrinter.cpp

```diff
@@ -4043,15 +4069,472 @@ void PrintAST::visitMissingMemberDecl(MissingMemberDecl *decl) {
+  visit(expr->getFn());
+  Printer << "(";
+  auto args = expr->getArgs()->getOriginalArgs();
+  bool isFirst = true;
+  // FIXME: handle trailing closures.
+  for (auto arg : *args) {
+    if (!isFirst) {
+      Printer << ", ";
+    }
+    printArgument(arg);
+    isFirst = false;
+  }
+  Printer << ")";
+}
+
+void PrintAST::printArgument(const Argument &arg) {
+  auto label = arg.getLabel();
+  if (!label.empty()) {
+    Printer << label.str();
+    Printer << ": ";
+  }
+  if (arg.isInOut()) {
+    Printer << "&";
+  }
+  visit(arg.getExpr());
+}
+
+void PrintAST::visitLoadExpr(LoadExpr *expr) {
+  visit(expr->getSubExpr());
+}
+
+void PrintAST::visitTypeExpr(TypeExpr *expr) {
+  printType(expr->getType());
+}
+
+void PrintAST::visitArrayExpr(ArrayExpr *expr) {
+  Printer << "[";
+  bool isFirst = true;
+  auto elements = expr->getElements();
+  for (auto element : elements) {
+    if (!isFirst) {
+      Printer << ", ";
+    }
+    visit(element);
+    isFirst = false;
+  }
+  Printer << "]";
+}
+
+void PrintAST::visitDictionaryExpr(DictionaryExpr *expr) {
+  Printer << "[";
+  bool isFirst = true;
+  auto elements = expr->getElements();
+  for (auto element : elements) {
+    auto *tupleExpr = cast<TupleExpr>(element);
+    if (!isFirst) {
+      Printer << ", ";
+    }
+    visit(tupleExpr->getElement(0));
+    Printer << ": ";
+    visit(tupleExpr->getElement(1));
+    isFirst = false;
+  }
+  Printer << "]";
+}
+
+void PrintAST::visitArrowExpr(ArrowExpr *expr) {
+}
+
+void PrintAST::visitAwaitExpr(AwaitExpr *expr) {
+  Printer << "await ";
+  visit(expr->getSubExpr());
+}
+
+void PrintAST::visitInOutExpr(InOutExpr *expr) {
+  visit(expr->getSubExpr());
+}
+
+void PrintAST::visitParenExpr(ParenExpr *expr) {
+  Printer << "(";
+  visit(expr->getSubExpr());
+  Printer << ")";
+}
+
+void PrintAST::visitTupleExpr(TupleExpr *expr) {
+  Printer << "(";
+  bool isFirst = true;
+  auto elements = expr->getElements();
+  for (auto element : elements) {
+    if (!isFirst) {
+      Printer << ", ";
+    }
+    visit(element);
+    isFirst = false;
+  }
+  Printer << ")";
+}
+
+void PrintAST::visitAssignExpr(AssignExpr *expr) {
+  visit(expr->getDest());
+  Printer << " = ";
+  visit(expr->getSrc());
+}
+
+void PrintAST::visitBinaryExpr(BinaryExpr *expr) {
+  visit(expr->getLHS());
+  Printer << " ";
+  if (auto operatorRef = expr->getFn()->getMemberOperatorRef()) {
+    Printer << operatorRef->getDecl()->getBaseName();
+  }
+  Printer << " ";
+  visit(expr->getRHS());
+}
+
+void PrintAST::visitCoerceExpr(CoerceExpr *expr) {
+}
+
+void PrintAST::visitOneWayExpr(OneWayExpr *expr) {
+}
+
+void PrintAST::visitClosureExpr(ClosureExpr *expr) {
+}
+
+void PrintAST::visitDeclRefExpr(DeclRefExpr *expr) {
+  Printer << expr->getDecl()->getBaseName();
+}
+
+void PrintAST::visitDotSelfExpr(DotSelfExpr *expr) {
+  visit(expr->getSubExpr());
+  Printer << ".self";
+}
+
+void PrintAST::visitErasureExpr(ErasureExpr *expr) {
+  visit(expr->getSubExpr());
+}
```

Comment:
I wonder if we could stamp out the same implementation for every `ImplicitConversionExpr` (as I don't believe any of them would require printing logic)? ExprNodes.def would probably have to be updated to use a new macro kind for such expressions, but then we could have e.g:

```cxx
#define IMPLICIT_CONVERSION_EXPR(Id, Parent) \
void PrintAST::visit##Id##Expr(Id *expr) {   \
  visit(expr->getSubExpr());                 \
}
#include "swift/AST/ExprNodes.def"
```

Doesn't have to be done in this PR though.

Reply:
I wonder if that's worth it for those one-liners. In any case, I don't want to attempt that in this PR.